Source/WebCore/dom/TrustedTypeUtils.cpp

```diff
@@ -0,0 +1,170 @@
+
+#include "ContentSecurityPolicy.h"
+#include "Document.h"
+#include "JSDOMExceptionHandling.h"
+#include "TrustedTypePolicy.h"
+#include "TrustedTypePolicyFactory.h"
+#include "WindowOrWorkerGlobalScopeTrustedTypes.h"
+#include "WorkerGlobalScope.h"
+#include <JavaScriptCore/HeapInlines.h>
+#include <JavaScriptCore/JSCInlines.h>
+#include <JavaScriptCore/JSCJSValueInlines.h>
+#include <JavaScriptCore/JSCast.h>
+
+namespace WebCore {
+using namespace JSC;
+
+struct TrustedTypeVisitor {
+    String operator()(std::nullptr_t)
+    {
+        return nullString();
+    }
+    String operator()(Ref<TrustedHTML> value)
+    {
+        return value->toString();
+    }
+    String operator()(Ref<TrustedScript> value)
+    {
+        return value->toString();
+    }
+    String operator()(Ref<TrustedScriptURL> value)
+    {
+        return value->toString();
+    }
+};
+
+String trustedTypeToString(const TrustedType trustedType)
+{
+    switch (trustedType) {
+    case TrustedType::TrustedHTML:
+        return "TrustedHTML"_s;
+    case TrustedType::TrustedScript:
+        return "TrustedScript"_s;
+    case TrustedType::TrustedScriptURL:
+        return "TrustedScriptURL"_s;
+    default:
+        ASSERT_NOT_REACHED();
+        return nullString();
+    }
+}
+
+// https://w3c.github.io/trusted-types/dist/spec/#process-value-with-a-default-policy-algorithm
+std::variant<std::nullptr_t, Ref<TrustedHTML>, Ref<TrustedScript>, Ref<TrustedScriptURL>> processValueWithDefaultPolicy(ScriptExecutionContext& scriptExecutionContext, const TrustedType expectedType, const String& input, const String& sink)
+{
+    VM& vm = scriptExecutionContext.vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    RefPtr<TrustedTypePolicy> protector = nullptr;
+    if (RefPtr document = dynamicDowncast<Document>(scriptExecutionContext)) {
+        if (auto window = document->domWindow()) {
+            auto trustedTypesFactory = WindowOrWorkerGlobalScopeTrustedTypes::trustedTypes(*window);
+            protector = trustedTypesFactory->defaultPolicy();
+        }
+    } else if (RefPtr workerGlobalScope = dynamicDowncast<WorkerGlobalScope>(scriptExecutionContext)) {
+        auto trustedTypesFactory = WindowOrWorkerGlobalScopeTrustedTypes::trustedTypes(*workerGlobalScope);
+        protector = trustedTypesFactory->defaultPolicy();
+    }
+
+    if (!protector)
+        return nullptr;
+
+    auto jsExpectedType = JSC::jsString(vm, trustedTypeToString(expectedType));
+    JSC::Strong<JSC::Unknown> strongExpectedType(vm, jsExpectedType);
+    auto jsSink = JSC::jsString(vm, sink);
+    JSC::Strong<JSC::Unknown> strongSink(vm, jsSink);
+    FixedVector<JSC::Strong<JSC::Unknown>> arguments({ strongExpectedType, strongSink });
+    auto policyValueHolder = protector->getPolicyValue(expectedType, input, WTFMove(arguments), false);
+    if (policyValueHolder.hasException()) {
+        propagateException(*scriptExecutionContext.globalObject(), scope, policyValueHolder.releaseException());
```

Comment:
I would tend to return the exception here and let the caller deal with throwing.
Hopefully that would be the `getTrustedTypeCompliantString` with global object version.

Reply:
I've rejigged this to hopefully work more how you'd like. I'm not sure how best to "throw" would be inside of JSDOMConvert, I've currently just called propagateException but in the other file instead?